### PR TITLE
[5.1] Nested Route Groups with as missing "in the middle"

### PR DESCRIPTION
**Pull request by ebynum**

### CODE REVIEW

**Comment by taylorotwell** [comment]
Doesn't need to be static. Needs doc block.

**Comment by ebynum** [comment]
I feel like an idiot for missing the doc block. Fixed.

**Comment by ebynum** [comment]
If `formatAsPrefix` doesn't need to be static, doesn't that mean that `protected static function formatGroupPrefix($new, $old)` and `protected static function formatUsesPrefix($new, $old)` shouldn't be static as well?

File: src/Illuminate/Routing/Router.php [hunk 1]
```diff
@@ -440,6 +438,24 @@ protected static function formatGroupPrefix($new, $old)
...
+     *
+     * @param  array  $new
+     * @param  array  $old
+     * @return string|null
+     */
+    protected static function formatAsPrefix($new, $old)
+    {
+        if (isset($new['as']) && isset($old['as'])) {
+            return $old['as'].$new['as'];
+        } elseif (isset($new['as'])) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Router.php]: don't need elseif, just start a new if statement please

File: src/Illuminate/Routing/Router.php [hunk 2]
```diff
@@ -440,6 +438,24 @@ protected static function formatGroupPrefix($new, $old)
...
+            return $new['as'];
+        }
+        
+        return isset($old['as']) ? $old['as'] : null;
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Router.php]:
```php
if (isset($old['as'])) {
   return $old['as'];
}
```

**Comment by lukasgeiter** [comment]
Wait, does this the same as my PR? #9211

**Comment by GrahamCampbell** [comment]
I've closed that duplicate.

**Comment by lukasgeiter** [comment]
@GrahamCampbell Thanks. However I feel like my implementation was way simpler. I don't really think this needs an extra method.

**Comment by ebynum** [comment]
@lukasgeiter, @GrahamCampbell I don't disagree that your implementation is "simpler". I just matched the existing convention from `formatUsesPrefix` and `formatAsPrefix`. Each of them could have been implemented the same way.

File: src/Illuminate/Routing/Router.php [hunk 3]
```diff
@@ -440,6 +439,25 @@ protected static function formatGroupPrefix($new, $old)
...
+     * @param  array  $new
+     * @param  array  $old
+     * @return string|null
+     */
+    protected static function formatAsPrefix($new, $old)
+    {
+        if (isset($new['as']) && isset($old['as'])) {
+            return $old['as'].$new['as'];
+        }
+        if (isset($new['as'])) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Router.php]: please leave a blank line before starting this if block :)

Comment (GrahamCampbell) [inline src/Illuminate/Routing/Router.php]: Actually, you shouldn't be doing the isset new check twice.

Comment (GrahamCampbell) [inline src/Illuminate/Routing/Router.php]:
```php
if (isset($new['as'])) {
    if (isset($old['as'])) {
        return $old['as'].$new['as'];
    }

    return $new['as'];
}

return isset($old['as']) ? $old['as'] : null;
```

Comment (lukasgeiter) [inline src/Illuminate/Routing/Router.php]: if (isset($new['as'])) {
	    return isset($old['as']) ? $old['as'].$new['as'] : $new['as'];
    }

    return isset($old['as']) ? $old['as'] : null;

**Comment by lukasgeiter** [comment]
>I just matched the existing convention from `formatUsesPrefix`

Good point, then the method indeed makes sense :)

**Comment by ebynum** [comment]
Cleaned up multi-if blocks on both `formatAsPrefix` and `formatUsesPrefix`.

File: src/Illuminate/Routing/Router.php (hunk 3 above)
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Router.php]: Yeh, that looks good. :)

**Comment by ebynum** [comment]
So @GrahamCampbell ... should I change the formatting to this style, or leave it as is?
```
        if (isset($new['as'])) {
            return isset($old['as'])
                ? $old['as'].$new['as']
                : $new['as'];
        }
```

**Comment by GrahamCampbell** [comment]
Yeh, changing it would be good, but put that ternary thing all on one line please.

**Comment by ebynum** [comment]
Put both ternary "blocks" on single lines.

File: tests/Routing/RoutingRouteTest.php [hunk 4]
```diff
@@ -549,18 +549,21 @@ public function testModelBindingWithBindingClosure()
     public function testGroupMerging()
     {
         $old = ['prefix' => 'foo/bar/'];
-        $this->assertEquals(['prefix' => 'foo/bar/baz', 'namespace' => null, 'where' => []], Router::mergeGroup(['prefix' => 'baz'], $old));
+        $this->assertEquals(['prefix' => 'foo/bar/baz', 'namespace' => null, 'as' => null, 'where' => []], Router::mergeGroup(['prefix' => 'baz'], $old));
```
Comment (taylorotwell) [inline tests/Routing/RoutingRouteTest.php]: Why did you change these tests? They shouldn't be changed.

Comment (ebynum) [inline tests/Routing/RoutingRouteTest.php]: I changed these tests to keep everything consistent. The `mergeGroup` method of Router "forced" `null` values for `prefix` and `namespace` when they weren't set manually. Everything else about the `as` option is consistent with them, and so I changed its structure in `mergeGroup` to match those two. Making it always show up (like they do) means that it gets a null when it hasn't been set at all, which means that calling `::mergeGroup()` will create a null value for `as`.

I added the `null`s for `as` just like the existing `null`s for `prefix` and `namespace`.

Comment (taylorotwell) [inline tests/Routing/RoutingRouteTest.php]: Change the tests back to how they were please.

**Comment by ebynum** [comment]
Switched to less complicated version originally suggested by @lukasgeiter ... it's now inconsistent with the `prefix` and `namespace` properties, but works with all existing tests and still fixes the bug.
